common-lib/src/cmr/common/memorydb/connection.clj

```diff
@@ -0,0 +1,88 @@
+(ns cmr.common.memorydb.connection
+  "Contains a record definition that implements the Lifecycle protocol. This
+  record is also intended to used for implementing ConceptStore, ConceptSearch,
+  and ProviderStore protocols.
+
+  This namespace was created to bring the MemoryStore in as close parity as
+  possible to the OracleStore in cmr.oracle.connection."
+  (:require
+   [cmr.common.lifecycle :as lifecycle]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Utility and support functions
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defn init-concepts
+  "Initialize the in-memory concepts tracker to be ordered for easy retrieval
+  of most recently added concept."
+  [opts]
+  (reverse (sort-by :revision-id (:concepts opts))))
+
+(defn init-next-id
+  [opts]
+  "Initialize the provided next-id."
+  (dec (:next-id opts)))
+
+(defn init-next-transaction-id
+  "Initialize the next-transaction-id to the given amount or set it to a
+  default of 0."
+  [opts]
+  (or (:next-transaction-id opts) 0))
+
+(defn init-providers
+  "Initialize the providers lookup table to the given hash map or set it to a
+  default of an empty hash map."
+  [opts]
+  (or (:providers opts) {}))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; A stoage record for use in parity with cmr.oracle.connection
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defrecord MemoryStore
```

Comment:
These are all required to be atoms in order to work with the code written against this record. It's not obvious in the code that later uses these that they need to be deref'ed. For example the old name of concepts-atom was more clear for a maintainer. I'm not sure what I'm suggesting to do, I think at a minimum documenting they must be atoms here would help - maybe changing the names to atom would help too?

Reply:
Oh I actually see they used to all be called -atom. We should put that change back to make it clear.

Reply:
Really? That's like naming all string vars something-string, floats something-float ...

The code that actually creates the instance generates the atoms, and all the references to the memory db's attributes use atom operations (`set!` and `@`), so it's clear everywhere that these are atoms ... maybe we need to have a chat so you can clarify for me what might be unclear to a maintainer? And maybe we can solve that with docs?

Reply:
Taking a closer look at this, yeah -- at the very least I'll update the comments in this record to indicate the need for these to be atom. We can chat about the rest of it more, too ...

Reply:
Okay, I've updated the record as well as its attributes with code comments. Also updated the constructor docstring.

Reply:
I don't see any benefit to removing -atom from the vars. I do see a benefit to having -atom in there. I think self documenting vars are better (and by naming them explicitly we don't need documentation).

It's the principle of least surprise (I might be making up that name), but when I see something called concepts I expect it to be a collection of concepts like everywhere else the name concepts is used in the code. When I see concepts-atom I think of an atom holding a collection of concepts.